### PR TITLE
New price sourcing

### PR DESCRIPTION
**Pull request by sambarboza**
# Description
Change source o exchange rates in order to have a more stable exchange rate source and standardize prices between Android, iOS and other available POS systems.

### Main source: 
Use Spark API. (https://api.get-spark.com/list) - https://github.com/kodaxx/DashRates-API.
One advantage of using their API is that we can get basically the same information that we have in our Android Wallet currently (before this PR) by doing just one api call instead of three. Also, they have a 30 seconds cache so BitcoinAvg and CryptoCompare APIs aren’t hit every time a price request is made.

### First Fallback:
(In case of Spark API’s downtime or not reachable by some regions):
Use the same method used by Spark API but making the requests directly from the device to each API. BTC/Fiat prices from BitcoinAvg API, Dash/BTC and Dash/VES price from CryptoCompare API. 

### Second Fallback
BTC/Fiat from BitPay API, VES/BTC from LocalBitcoins, Dash/BTC from Poloniex.

# Changes
- Change main exchange rate source to https://api.get-spark.com/
- Added 2 fallbacks to main exchange rate source.
- Removed `ExchangeRatesProvider` and `ExchangeRatesLoader` classes.
- Using Retrofit to get exchange rate from web APIs.
- Using Room to cache data.
- Implemented repository pattern.
- Using LiveData to automatically update UI when new exchange rates are loaded.

### CODE REVIEW

**Comment by HashEngineering** [comment]
@sambarboza 

I get this error when running on an Android 9.0 emulator and an Android 8.0 device as built on a windows machine.

This came up in a search for this exception:https://stackoverflow.com/questions/49038667/runtimeexceptionclassnotfoundexception-android-arch-lifecycle-processlifecycleo


```
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: hashengineering.dash.wallet.beta, PID: 21147
    java.lang.RuntimeException: Unable to get provider android.arch.lifecycle.ProcessLifecycleOwnerInitializer: java.lang.ClassNotFoundException: Didn't find class "android.arch.lifecycle.ProcessLifecycleOwnerInitializer" on path: DexPathList[[zip file "/data/app/hashengineering.dash.wallet.beta-1mcBdtOP6XGs-QgpyGZxxg==/base.apk", zip file "/data/app/hashengineering.dash.wallet.beta-1mcBdtOP6XGs-QgpyGZxxg==/split_lib_directories_apk.apk", zip file "/data/app/hashengineering.dash.wallet.beta-1mcBdtOP6XGs-QgpyGZxxg==/split_lib_resources_apk.apk"],nativeLibraryDirectories=[/data/app/hashengineering.dash.wallet.beta-1mcBdtOP6XGs-QgpyGZxxg==/lib/arm, /data/app/hashengineering.dash.wallet.beta-1mcBdtOP6XGs-QgpyGZxxg==/base.apk!/lib/armeabi-v7a, /data/app/hashengineering.dash.wallet.beta-1mcBdtOP6XGs-QgpyGZxxg==/split_lib_directories_apk.apk!/lib/armeabi-v7a, /data/app/hashengineering.dash.wallet.beta-1mcBdtOP6XGs-QgpyGZxxg==/split_lib_resources_apk.apk!/lib/armeabi-v7a, /system/lib, /vendor/lib]]
        at android.app.ActivityThread.installProvider(ActivityThread.java:6575)
        at android.app.ActivityThread.installContentProviders(ActivityThread.java:6127)
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6037)
        at android.app.ActivityThread.-wrap1(Unknown Source:0)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1764)
        at android.os.Handler.dispatchMessage(Handler.java:105)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6938)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:327)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1374)
     Caused by: java.lang.ClassNotFoundException: Didn't find class "android.arch.lifecycle.ProcessLifecycleOwnerInitializer" on path: DexPathList[[zip file "/data/app/hashengineering.dash.wallet.beta-1mcBdtOP6XGs-QgpyGZxxg==/base.apk", zip file "/data/app/hashengineering.dash.wallet.beta-1mcBdtOP6XGs-QgpyGZxxg==/split_lib_directories_apk.apk", zip file "/data/app/hashengineering.dash.wallet.beta-1mcBdtOP6XGs-QgpyGZxxg==/split_lib_resources_apk.apk"],nativeLibraryDirectories=[/data/app/hashengineering.dash.wallet.beta-1mcBdtOP6XGs-QgpyGZxxg==/lib/arm, /data/app/hashengineering.dash.wallet.beta-1mcBdtOP6XGs-QgpyGZxxg==/base.apk!/lib/armeabi-v7a, /data/app/hashengineering.dash.wallet.beta-1mcBdtOP6XGs-QgpyGZxxg==/split_lib_directories_apk.apk!/lib/armeabi-v7a, /data/app/hashengineering.dash.wallet.beta-1mcBdtOP6XGs-QgpyGZxxg==/split_lib_resources_apk.apk!/lib/armeabi-v7a, /system/lib, /vendor/lib]]
        at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:93)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:379)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
        at android.app.ActivityThread.installProvider(ActivityThread.java:6560)
        at android.app.ActivityThread.installContentProviders(ActivityThread.java:6127) 
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6037) 
        at android.app.ActivityThread.-wrap1(Unknown Source:0) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1764) 
        at android.os.Handler.dispatchMessage(Handler.java:105) 
        at android.os.Looper.loop(Looper.java:164) 
        at android.app.ActivityThread.main(ActivityThread.java:6938) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:327) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1374) 
    	Suppressed: java.io.IOException: No original dex files found for dex location (arm) /data/app/hashengineering.dash.wallet.beta-1mcBdtOP6XGs-QgpyGZxxg==/split_lib_resources_apk.apk
        at dalvik.system.DexFile.openDexFileNative(Native Method)
        at dalvik.system.DexFile.openDexFile(DexFile.java:353)
        at dalvik.system.DexFile.<init>(DexFile.java:100)
        at dalvik.system.DexFile.<init>(DexFile.java:74)
        at dalvik.system.DexPathList.loadDexFile(DexPathList.java:374)
        at dalvik.system.DexPathList.makeDexElements(DexPathList.java:337)
        at dalvik.system.DexPathList.<init>(DexPathList.java:157)
        at dalvik.system.BaseDexClassLoader.<init>(BaseDexClassLoader.java:65)
        at dalvik.system.PathClassLoader.<init>(PathClassLoader.java:64)
        at com.android.internal.os.PathClassLoaderFactory.createClassLoader(PathClassLoaderFactory.java:43)
        at android.app.ApplicationLoaders.getClassLoader(ApplicationLoaders.java:69)
        at android.app.ApplicationLoaders.getClassLoader(ApplicationLoaders.java:36)
E/AndroidRuntime:     at android.app.LoadedApk.createOrUpdateClassLoaderLocked(LoadedApk.java:676)
        at android.app.LoadedApk.getClassLoader(LoadedApk.java:709)
        at android.app.LoadedApk.getResources(LoadedApk.java:942)
        at android.app.ContextImpl.createAppContext(ContextImpl.java:2303)
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:5937)

```

**Comment by tomasz-ludek** [comment]
@HashEngineering ,
you probably just need to disable *Instant Run* in AS settings.

**Comment by HashEngineering** [comment]
Instant Run is now disabled and the app runs.  Thank you.

**Comment by HashEngineering** [comment]
Since building and running worked, the next thing to do was test each of the three methods of getting the Dash price.  My method of testing was to run the app without modification and check for the success or failure of the first method.  And it worked:

```
2019-01-08 16:49:43.532 23305-23399/hashengineering.dash.wallet.beta I/ExchangeRatesRepository: [pool-1-thread-1] 
	exchange rates updated successfully with de.schildbach.wallet.rates.DashRatesClient@f160b67
```
Then I commented out the first method and the second method would be used.  It worked.

```
2019-01-08 16:52:24.439 25792-25922/hashengineering.dash.wallet.beta I/ExchangeRatesRepository: [pool-1-thread-1] 
	exchange rates updated successfully with de.schildbach.wallet.rates.DashRatesFirstFallback@aa3a50d
```

Lastly, the first and second methods were commented out and the last method failed on Android 9.

```
2019-01-08 16:46:48.083 16023-16154/hashengineering.dash.wallet.beta E/ExchangeRatesRepository: [pool-1-thread-1] 
	failed to fetch exchange rates with de.schildbach.wallet.rates.DashRatesSecondFallback@fbedbbe
    java.net.UnknownServiceException: CLEARTEXT communication to dash.casa not permitted by network security policy
    	at okhttp3.internal.connection.RealConnection.connect(RealConnection.java:139) ~[na:0.0]
    	at okhttp3.internal.connection.StreamAllocation.findConnection(StreamAllocation.java:192) ~[na:0.0]
    	at okhttp3.internal.connection.StreamAllocation.findHealthyConnection(StreamAllocation.java:121) ~[na:0.0]
    	at okhttp3.internal.connection.StreamAllocation.newStream(StreamAllocation.java:100) ~[na:0.0]
    	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:42) ~[na:0.0]
    	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92) ~[na:0.0]
    	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67) ~[na:0.0]
    	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93) ~[na:0.0]
    	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92) ~[na:0.0]
    	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67) ~[na:0.0]
    	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93) ~[na:0.0]
    	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92) ~[na:0.0]
    	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:120) ~[na:0.0]
    	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92) ~[na:0.0]
    	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67) ~[na:0.0]
    	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:185) ~[na:0.0]
    	at okhttp3.RealCall.execute(RealCall.java:69) ~[na:0.0]
    	at retrofit2.OkHttpCall.execute(OkHttpCall.java:180) ~[na:0.0]
    	at retrofit2.ExecutorCallAdapterFactory$ExecutorCallbackCall.execute(ExecutorCallAdapterFactory.java:91) ~[na:0.0]
    	at de.schildbach.wallet.rates.DashCasaClient.getRates(DashCasaClient.java:38) ~[na:0.0]
    	at de.schildbach.wallet.rates.DashRatesSecondFallback.getRates(DashRatesSecondFallback.java:36) ~[na:0.0]
    	at de.schildbach.wallet.rates.ExchangeRatesRepository$1.run(ExchangeRatesRepository.java:84) ~[na:0.0]
    	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167) ~[na:0.0]
    	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) ~[na:0.0]
    	at java.lang.Thread.run(Thread.java:764) ~[na:0.0]
```
		
This exception happens here:
Line 36 of DashRatesSecondFallback.java:
`	        BigDecimal dashVesPrice = DashCasaClient.getInstance().getRates().body().getDashVesPrice();`

Link to some information about this exception, as ClearText is not supported by Android 9 and I am running on an Android 9 emulator	:		
			https://developer.android.com/training/articles/security-config#CleartextTrafficPermitted

**Comment by HashEngineering** [comment]
I will report again when I try the failed method from Android 9 on an Android 8 device.

**Comment by HashEngineering** [comment]
On Android 8, the DashRatesSecondFallback method is successful.
```
2019-01-08 18:54:56.421 6027-6053/hashengineering.dash.wallet.beta I/ExchangeRatesRepository: [pool-1-thread-1] 
exchange rates updated successfully with de.schildbach.wallet.rates.DashRatesSecondFallback@b7f8ad8
```

We just need to fix the case on Android 9.

**Comment by sambarboza** [comment]
@HashEngineering, thank you!
This commit should fix the issue on Android 9: https://github.com/HashEngineering/dash-wallet/pull/159/commits/a419fa10d636230e33841bc821235e8e5cb4e1c8

**Comment by HashEngineering** [comment]
Verified that this last commit fixed the issue on Android 9:

```
2019-01-08 17:12:26.147 15334-15395/hashengineering.dash.wallet.beta I/ExchangeRatesRepository: [pool-1-thread-1] 
exchange rates updated successfully with de.schildbach.wallet.rates.DashRatesSecondFallback@96d91a0
```

File: wallet/src/de/schildbach/wallet/ui/ExchangeRatesFragment.java [hunk 1]
```diff
@@ -112,7 +113,29 @@ public void onCreate(final Bundle savedInstanceState) {
...
+        exchangeRatesViewModel.isLoading().observe(this, new Observer<Boolean>() {
+            @Override
+            public void onChanged(Boolean isLoading) {
+                loading.setVisibility(Boolean.TRUE.equals(isLoading) ? View.VISIBLE : View.GONE);
+            }
+        });
+        exchangeRatesViewModel.hasError().observe(this, new Observer<Boolean>() {
+            @Override
+            public void onChanged(@android.support.annotation.Nullable Boolean aBoolean) {
+                if (Boolean.TRUE.equals(true)) {
```
[the remark this answers is not in the document]
Comment (sambarboza) [reply wallet/src/de/schildbach/wallet/ui/ExchangeRatesFragment.java]: Good catch, thank you! Fixed.

**Pull request merged**